### PR TITLE
Edit basic sheets and create Organisation/Processes with Substanced Admin Interface (sdi)

### PR DESCRIPTION
**Pull request by joka**
This pull request depends on #2259 !

SDI Extensions:
----------------------

If you go to the sid as god ( /manage ) you have to new features:

* edit basic sheets
* create processes and organisations

Backend internal Changes:
-----------------------------------

* Fix request attribute of IPermisison sheet might be none
* only store sheet appstruct data if values have changed
* allow to set workflow state to current state (does no trigger transition)
* preserve sheet order

API Changes:
------------------

* field name of IName sheets is required (post only)

### CODE REVIEW

File: src/adhocracy_core/adhocracy_core/sheets/__init__.py [hunk 1]
```diff
@@ -192,7 +193,13 @@ def set(self,
...
             self.registry.notify(event)
         return bool(appstruct)
-        # TODO: only store struct if values have changed
+
+    def _filter_unchanged_data(self, new: dict, old: dict) -> dict:
+        changed = {}
+        for key, value in new.items():
+            if value != old[key]:
+                changed[key] = value
+        return changed
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/__init__.py]: What about:

```
diff = set(new.items()) - set(old.items())
changed = dict((k, v) for k, v in diff)
return changed
```

?

File: src/adhocracy_core/adhocracy_core/sdi/views/sheets.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+from substanced.util import _
+
+from adhocracy_core.interfaces import IResource
+from adhocracy_core.interfaces import IResourceSheet
+
+
+class FormView(SDFormView):
+    """View class that autogenerates forms based on :term:`schema` .
+
+    I uses :mod:`colander` schema to define the data structure and
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sdi/views/sheets.py]: Can we avoid the first person for documenting? It sounds strange.

File: src/adhocracy_core/adhocracy_core/rest/test_views.py [hunk 3]
```diff
@@ -910,7 +910,7 @@ def make_one(self, request, context):
 
     def test_post_without_token_authentication_policy(self, request, context):
         inst = self.make_one(context, request)
-        with pytest.raises(IndexError):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/rest/test_views.py]: Commit message is not related to the change?

**Comment by pallix** [comment]
LGTM, modulo the comments above. Also it has already conflicts with master.

File: src/adhocracy_core/adhocracy_core/sheets/__init__.py (hunk 1 above)
Comment (joka) [inline src/adhocracy_core/adhocracy_core/sheets/__init__.py]: yes thats nicer, but it does work with list values

File: src/adhocracy_core/adhocracy_core/sdi/views/contents.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    def sdi_addable_content(self):
+        registry = self.request.registry
+        introspector = registry.introspector
+        cts = []
+        meta_addable = self.request.registry.content\
+            .get_resources_meta_addable(self.context, self.request)
+        content_types_addable = [m.iresource.__identifier__ for m
+                                 in meta_addable]
+        for data in introspector.get_category('substance d content types'):
```
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sdi/views/contents.py]: what the purpose of getting all types of the system and then check if they are member of the addable set? Can't we just return the set?

Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sdi/views/contents.py]: Also why is the `is_sdi_addable` boolean not used to filter the addable resources?

File: src/adhocracy_core/adhocracy_core/sheets/__init__.py (hunk 1 above)
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sheets/__init__.py]: you mean it does not work when the value is a list?

File: src/adhocracy_core/adhocracy_core/sdi/views/contents.py (hunk 4 above)
Comment (pallix) [inline src/adhocracy_core/adhocracy_core/sdi/views/contents.py]: @joka ping!

File: src/adhocracy_core/adhocracy_core/sheets/__init__.py (hunk 1 above)
Comment (joka) [inline src/adhocracy_core/adhocracy_core/sheets/__init__.py]: > you mean it does not work when the value is a list?

yes, lists are not immuatable

File: src/adhocracy_core/adhocracy_core/sdi/views/contents.py (hunk 4 above)
Comment (joka) [inline src/adhocracy_core/adhocracy_core/sdi/views/contents.py]: > what the purpose of getting all types of the system and then check if they are member of the addable set? Can't we just return the set?

Looks like,  I have check this

Comment (joka) [inline src/adhocracy_core/adhocracy_core/sdi/views/contents.py]: > Also why is the is_sdi_addable boolean not used to filter the addable resources?

I used the substanced filter mechanism, it only lists resource types that have an AddView for the current context. But you are right, we can also just filter by is_sdi_addable.

**Comment by joka** [comment]
lgtm
